### PR TITLE
[#22] Add Dropper to decode SscPayload/Proof

### PR DESCRIPTION
**Pull request by ruhatch**
Add the `Dropper s` type to decode CBOR, throwing away the results. This allows us to traverse the structure of the old blockchain, ignoring Haskell structure that we no longer care about.

### CODE REVIEW

File: binary/src/Cardano/Binary/Class/Drop.hs [hunk 1]
```diff
@@ -0,0 +1,59 @@
+-- | A 'Dropper s' is a 'Decoder s ()', that is a decoder that returns nothing
```
Comment (dnadales) [inline binary/src/Cardano/Binary/Class/Drop.hs:1]: "that is, a decoder ..." (missing comma?)

File: binary/src/Cardano/Binary/Class/Drop.hs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+       , dropSet
+       , dropTuple
+       , dropTriple
+       , dropWord64
+       ) where
+
+import           Cardano.Prelude
+
+import qualified Codec.CBOR.Decoding as D
+
```
Comment (dnadales) [inline binary/src/Cardano/Binary/Class/Drop.hs:21]: Extra space?

File: binary/src/Cardano/Binary/Class/Drop.hs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+dropBytes :: Dropper s
+dropBytes = void D.decodeBytesCanonical
+
+-- | Drop a list of values using the supplied `Dropper` for each element
+dropList :: Dropper s -> Dropper s
+dropList dropElems = do
+  D.decodeListLenIndef
+  D.decodeSequenceLenIndef const () id dropElems
+
+dropMap :: Dropper s -> Dropper s -> Dropper s
```
Comment (dnadales) [inline binary/src/Cardano/Binary/Class/Drop.hs:34]: It seems obvious what `dropMap`, `dropSet`, etc do, but I was wondering whether we need to put the comments, just for the sake of consistency. I'm fine with not having them though.

File: binary/src/Cardano/Binary/Class/Primitive.hs [hunk 4]
```diff
@@ -112,12 +109,12 @@ serializeWith firstChunk nextChunk =
 -- | Deserialize a Haskell value from the external binary representation
 --   (which must have been made using 'serialize' or related function).
 --
---   /Throws/: @'CBOR.Read.DeserialiseFailure'@ if the given external
+--   /Throws/: @'Read.DeserialiseFailure'@ if the given external
```
Comment (dnadales) [inline binary/src/Cardano/Binary/Class/Primitive.hs:112]: Hmmm, do we want to make the error explicit in the types somehow, or are we ok with this?

File: binary/src/Cardano/Binary/Class/Primitive.hs [hunk 5]
```diff
@@ -140,56 +137,42 @@ deserialize' = deserialize . BSL.fromStrict
...
+  :: (forall s . D.Decoder s a)
+  -> BSL.ByteString
+  -> Either (Read.DeserialiseFailure, BS.ByteString) (a, BS.ByteString)
+deserialiseDecoder decoder bs0 =
+  runST (supplyAllInput bs0 =<< Read.deserialiseIncremental decoder)
+
+supplyAllInput
+  :: BSL.ByteString
+  -> Read.IDecode s a
+  -> ST
```
Comment (dnadales) [inline binary/src/Cardano/Binary/Class/Primitive.hs:167]: This is a weird indentation...

File: crypto/src/Cardano/Crypto/SecretSharing.hs [hunk 6]
```diff
@@ -116,20 +126,29 @@ newtype Secret = Secret
...
+
 -- | Encrypted share which needs to be decrypted using 'VssKeyPair' first.
 newtype EncShare = EncShare
   { getEncShareVal :: Scrape.EncryptedSi
   } deriving (Show, Eq)
 
 deriving instance Bi EncShare
 
+dropEncShare :: Dropper s
+dropEncShare = dropBytes
```
Comment (dnadales) [inline crypto/src/Cardano/Crypto/SecretSharing.hs:150]: We have this pattern in 4 places, why not using `dropBytes` directly?

File: crypto/test/Test/Cardano/Crypto/Bi.hs [hunk 7]
```diff
@@ -217,69 +228,77 @@ roundTripProxyCertAeson = eachOf 100 genUnitProxyCert roundTripsAesonBuildable
...
 golden_DecShare = goldenTestBi decShare "test/golden/DecShare"
-  where
-    Just decShare = case sharedSecretData of
-        (_, _, xs) -> deterministic "ds" . uncurry decryptShare <$> fmap fst (uncons xs)
+ where
+  Just decShare = case sharedSecretData of
+    (_, _, xs) ->
+      deterministic "ds" . uncurry decryptShare <$> fmap fst (uncons xs)
+
+golden_legacy_DecShare :: Property
```
Comment (dnadales) [inline crypto/test/Test/Cardano/Crypto/Bi.hs:299]: I'm curious about when do we use underscore and when camel-case. Maybe this was inherited from QuickCheck props, but it'd be nice to use only one style.

**Review by dnadales — APPROVED** [review]

File: binary/src/Cardano/Binary/Class/Primitive.hs (hunk 4 above)
Comment (ruhatch) [reply binary/src/Cardano/Binary/Class/Primitive.hs:112]: This is currently only used in a couple of tests - I imagine we'll remove it entirely as it's a bit weird

File: binary/src/Cardano/Binary/Class/Drop.hs (hunk 2 above)
Comment (ruhatch) [reply binary/src/Cardano/Binary/Class/Drop.hs:21]: Nope - have been doing two NLs between imports and code start :)

File: binary/src/Cardano/Binary/Class/Primitive.hs (hunk 5 above)
Comment (ruhatch) [reply binary/src/Cardano/Binary/Class/Primitive.hs:167]: Yeah, not sure how I would make it better though - any suggestions?

File: crypto/src/Cardano/Crypto/SecretSharing.hs (hunk 6 above)
Comment (ruhatch) [reply crypto/src/Cardano/Crypto/SecretSharing.hs:150]: I ended up dropping it in a later PR :)

File: crypto/test/Test/Cardano/Crypto/Bi.hs (hunk 7 above)
Comment (ruhatch) [reply crypto/test/Test/Cardano/Crypto/Bi.hs:299]: You're right - I'll make an issue to change these across the whole codebase

File: binary/src/Cardano/Binary/Class/Primitive.hs (hunk 5 above)
Comment (dnadales) [reply binary/src/Cardano/Binary/Class/Primitive.hs:167]: Well, maybe in some cases we don't have to be that strict about the 80 chars limit :) That's the only suggestion I can think of...

**Pull request merged**